docs/advanced_topics.rst

```diff
@@ -27,10 +27,15 @@ this will set thread 0 to CPU 3, thread 1 to CPU 5, thread 2 to CPU 6, thread 3
 Memory consumption
 ------------------
 
-In DALI architecture all GPU operators process the whole batch at a time while the CPU operator processes one sample at a time.
-With DALI the memory is only growing, never shrinking. Hence each GPU buffer is as large as the largest possible batch, while the CPU buffers are as large as batch size multiplied by the size of the largest sample. Note that even though the CPU processes one sample at a time per thread, a vector of samples need to reside in the memory.
-This lazy allocation strategy reduces the number of total memory operations, and hence helps in cases when the memory allocation is expensive, like the pinned CPU memory or GPU memory.
-This is most visible for the operators whose output size may differ from sample to sample and from run to run. Operator with the fixed size outputs, such as crop, does not influence the overall memory consumption growth over time
+DALI uses three kinds of memory: host, host page-locked (pinned) and GPU.
+
+GPU and pinned memory allocation and freeing require device synchronization. For this reason, DALI avoids reallocating these kinds of memory whenever possible. The buffers allocated with this kind of storage will only grow when the existing buffer is too small to accommodate the requested shape. This allocation strategy reduces the number of total memory management operations and greatly increases the processing speed after the allocations have stabilized.
+
+In contrast, ordinary host memory is relatively cheap to allocate and free. To reduce host memory consumption, the buffers may shrink if the new requested size is smaller than a specified fraction of the old size (called shrink threshold). It can be adjusted to any value between 0 and 1, with the default being 0.9. The value can be controlled either via environment variable `DALI_HOST_BUFFER_SHRINK_THRESHOLD` or set in Python using `nvidia.dali.backend.SetHostBufferShrinkThreshold` function.
```

Comment:
Maybe add a sentence that `1` would mean to shrink always, and `0` would mean to never shrink.

Reply:
Done